Documentation/coding-guidelines/interop-pinvokes.md

```diff
@@ -218,6 +225,64 @@ Blittable structs are much more performant as they they can simply be used direc
 
 Pointers to structs in definitions must either be passed by `ref` or use `unsafe` and `*`.
 
+The compiler or marshaler will make sure any pointers in your struct are correctly aligned. For this reason seemingly "wrong" declarations can still work. For example PROCESS_BASIC_INFORMATION is documented as:
+
+```C
+typedef struct _PROCESS_BASIC_INFORMATION {
+    PVOID Reserved1;
+    PPEB PebBaseAddress;
+    PVOID Reserved2[2];
+    ULONG_PTR UniqueProcessId;
+    PVOID Reserved3;
+} PROCESS_BASIC_INFORMATION;
+```
+
+All the members of this struct have the underlying type of pointer, so a correct managed representation could be:
+
+```c#
+    struct PROCESS_BASIC_INFORMATION
+    {
+        public IntPtr Reserved1; // or UIntPtr
+        public IntPtr PebBaseAddress;
+        public IntPtr Reserved2a;
+        public IntPtr Reserved2b;
+        public IntPtr BasePriority;
+        public UIntPtr UniqueProcessId;
+        public UIntPtr Reserved3;
+    }
+```
+
+Under the covers, the first field in this struct is actually type `LONG` which would normally marshal as a 32-bit `int` in managed code. The reason that `IntPtr` still works is that the next field happens to be a pointer, which must be aligned. To
```

Comment:
I do not think that this is relevant here. This mismatch is a cosmetic bug in Windows SDK. Not something one needs to be worried about when writing PInvokes.

Reply:
This should just say "We always prefer to match the managed struct as closely as possible to the shape and names that is used in the documentation or the public header files."

Reply:
Right, I don't care about "cosmetic bugs" in Windows headers. The reason I mention this is that sometimes you are confronted with different versions of the native struct. Eg., this one is around the web with NTSTATUS. I like to understand what I am doing rather than blindly pasting, and it took me a bit of time (starting with confusion about LONG being fixed at 32 bits) before I realized it worked because of padding.

If you like I could remove this example and just mention that sometimes you can see IntPtr used in place of a shorter type, and it works because the subsequent field is a real pointer.

Reply:
Similar mismatches occur occasionally in many different shapes. For example, `long` used instead of `double`, `void*` used instead of `va_list/ArgIterator`, ... .

You always need a deep low-level understanding to figure out why it happens to work; or why it happens to work on one platform and not on the other platform that has different alignment rules or calling convention.

Reply:
OK I deleted most of this down to a simple example of a case of the general point: follow official documentation.

As for deep low-level understanding - I hope whenever a mistake is common, we record it here. It's better than me pointing folks to my OneNote.

Reply:
Alright I deleted the example. I think it would be worth saying something about padding and alignment in general. We don't say whether we require `LayoutKind.Sequential` even though it's the C# default, and we aren't consistent.

Reply:
(But I won't do that here)

Reply:
> We don't say whether we require `LayoutKind.Sequential` even though it's the C# default, and we aren't consistent.

Less ceremony and depending on defaults is generally better. Explicit `LayoutKind.Sequential` is not a no-op though. It disables unused field warnings as side-effect. (The alternative would be to disable the warning using `#pragma` or command line option.)